pkg/apps/application-list.jsx

```diff
@@ -28,18 +28,73 @@ import { Stack, StackItem } from "@patternfly/react-core/dist/esm/layouts/Stack/
 
 import { RebootingIcon } from "@patternfly/react-icons";
 
-import { is_installed } from "packagekit";
 import { get_manifest_config_matchlist } from "utils";
 import { read_os_release } from "os-release";
 import { EmptyStatePanel } from "cockpit-components-empty-state.jsx";
 import { useInit } from "hooks";
 
-import * as PackageKit from "./packagekit.js";
-import { icon_url, show_error, launch, ProgressBar, CancelButton } from "./utils";
+import { ProgressReporter, icon_url, show_error, launch, ProgressBar, CancelButton } from "./utils";
 import { ActionButton } from "./application.jsx";
+import { getPackageManager } from "packagemanager.js";
 
 const _ = cockpit.gettext;
 
+async function refresh_appstream_metadata(origin_files, config_packages, data_packages, progress_cb) {
+    /* In addition to refreshing the repository metadata, we also
+     * update all packages that contain AppStream collection metadata.
+     *
+     * AppStream collection metadata is arguably part of the
+     * repository metadata and should be updated during a regular
+     * refresh of the repositories.  On some distributions this is
+     * what happens, but on others (such as Fedora), the collection
+     * metadata is delivered in packages.  We find them and update
+     * them explicitly.
+     *
+     * Also, we have two explicit lists of packages, and we make sure
+     * that they are installed.  The first list contains packages that
+     * configure the system to retrieve AppStream data as part of
+     * repository metadata, and the second list contains packages that
+     * contain AppStream data themselves.
+     */
+    const packagemanager = await getPackageManager();
+    const progress = new ProgressReporter(0, 1, progress_cb);
+
+    if (config_packages.length > 0) {
+        progress.base = 0;
+        progress.range = 5;
+
+        await packagemanager.install_packages(config_packages, progress.progress_reporter);
+    }
+
+    const origin_pkgs = new Set(await packagemanager.find_file_packages(origin_files, progress.progress_reporter));
+
+    progress.base = 6;
+    progress.range = 69;
+    await packagemanager.refresh(true, progress.progress_reporter);
+
+    progress.base = 75;
+    progress.range = 5;
+
+    const updates = await packagemanager.get_updates(false, progress.progress_reporter);
+    const filtered_updates = [];
+
+    for (const update of updates) {
+        if (origin_pkgs.has(update.name))
+            filtered_updates.push(update);
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22611-295a8796-20251126-172456-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)